DDUF documentation (#1526)
huggingface/hub-docs: docs/hub/dduf.md

Author: yeah, there is a start at the very bottom of the page (could be moved to its own section)

# DDUF

> ## Overview

Reviewer: ```suggestion
<div class="flex justify-center">
    <img class="block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/DDUF/DDUF-Banner.svg"/>
</div>

## Overview
```

DDUF (Diffusion ??? Universal Format) is a file format designed to make storing, distributing, and using diffusion models much easier. Built on the ZIP file format, DDUF offers a standardized, efficient, and flexible way to package all parts of a diffusion model into a single, easy-to-manage file.

This work draws inspiration from the [GGUF](https://github.com/ggerganov/ggml/blob/master/docs/gguf.md) format.

> 

Reviewer 2: ```suggestion

We've seeded some DDUFs of popular formats for the community to play with: https://huggingface.co/DDUF, check them out!
```

<Tip>

We welcome contributions with open arms!

To create a widely adopted file format, we need early feedback from the community. Nothing is set in stone, and we value everyone's input. Is your use case not covered? Please let us know! Discussions about the DDUF format happen in the [DDUF organization](https://huggingface.co/spaces/DDUF/README/discussions/2).

</Tip>

## Motivation

Yet, another file format? Yes, but for good reasons!

> The primary goal of DDUF is to create a community-endorsed file format for diffusion models. Current model distribution methods often involve multiple separate files, different weight-saving formats, and managing files from various locations. DDUF aims to solve these challenges by packaging all model components into a single file, enforcing a consistent structure while being opinionated about saving formats.

Reviewer: ```suggestion
The primary goal of DDUF is to create a community-endorsed single-file format for diffusion models. Current model distribution methods often involve multiple separate files, different weight-saving formats, and managing files from various locations. DDUF aims to solve these challenges by packaging all model components into a single file, enforcing a consistent structure while being opinionated about saving formats.
```

## Key Features

1. **Single file** packaging.
2. Based on **ZIP file format** to leverage existing tooling.
3. No compression, ensuring **`mmap` compatibility** for fast loading and saving.
4. **HTTP-friendly**: metadata and file structure can be fetched remotely using HTTP Range requests.
5. **Flexible**: each model component is stored in its own directory, following the current `diffusers` structure.
6. **Safe**: uses `safetensors` as weights saving format and prohibits nested directories to prevent ZIP-bombs.

## Technical specifications

Technically, a `.dduf` file **is** a [`.zip` archive](https://en.wikipedia.org/wiki/ZIP_(file_format)). By building on a universally supported file format, we ensure robust tooling already exists. However, some constraints are enforced to meet diffusion models' requirements:
- Data must be stored uncompressed (flag `0`), allowing mmap-compatibility.

> - Data must be stored using ZIP64 protocol, enabling saving files above 2GB.

Reviewer 3: I think the original zip format has a maximum file size limit of 2^32 bytes = 4GB
```suggestion
- Data must be stored using ZIP64 protocol, enabling saving files above 4GB.
```

Author: Indeed! Thanks :)

- The archive can only contain `.json`, `.safetensors`, `.model` and `.txt` files.
- A `model_index.json` file must be present at the root of the archive. It must contain a key-value mapping with metadata about the model and its components.
- Each component must be stored in its own directory (e.g., `vae/`, `text_encoder/`). Nested files must use UNIX-style path separators (`/`).
- Each directory must correspond to a component in the `model_index.json` index.
- Each directory must contain a json config file (one of `config.json`, `tokenizer_config.json`, `image_processor.json`).
- Sub-directories are forbidden.

> Want to check if your file is valid? Check it out using this Space: https://huggingface.co/spaces/DDUF/dduf-check.

Reviewer 3: seems like the space is stuck and flagged as abusive 
![Screenshot 2024-12-12 at 14 39 34](https://github.com/user-attachments/assets/8b5f66f7-c7a3-40b3-81ef-2209e9e01a37)

> Want to check if your file is valid? Check it out using this Space: https://huggingface.co/spaces/DDUF/dduf-check.

Reviewer 4: removed the abuse flag

## Usage

The `huggingface_hub` provides tooling to handle DDUF files in Python. It includes built-in rules to validate file integrity and helpers to read and export DDUF files. The goal is to see this tooling adopted in the Python ecosystem, such as in the `diffusers` integration. Similar tooling can be developed for other languages (JavaScript, Rust, C++, etc.).

### How to read a DDUF file?

> Reading a DDUF file is as simple as calling `read_dduf_file` and passing a path as argument. Only the metadata is read, meaning this is a lightweight call that will not explode your memory. In the example below, we consider that you've already downloaded the [`FLUX.1-dev.dduf`](https://huggingface.co/DDUF/FLUX.1-dev-DDUF/blob/main/FLUX.1-dev.dduf) file locally.

Reviewer: ```suggestion
Reading a DDUF file is as simple as calling `read_dduf_file` and passing a path as argument. Only the metadata is read, meaning this is a lightweight call that will not make your memory explode. In the example below, we consider that you've already downloaded the [`FLUX.1-dev.dduf`](https://huggingface.co/DDUF/FLUX.1-dev-DDUF/blob/main/FLUX.1-dev.dduf) file locally.
```

...     yield "vae/config.json", pipe.vae.to_json_string().encode()
...     yield "vae/diffusion_pytorch_model.safetensors", safetensors.torch.save(pipe.vae.state_dict())
...     yield "text_encoder/config.json", pipe.text_encoder.config.to_json_string().encode()
...     yield "text_encoder/model.safetensors", safetensors.torch.save(pipe.text_encoder.state_dict())
...     # ... add more entries here

>>> export_entries_as_dduf(dduf_path="my-cool-diffusion-model.dduf", entries=as_entries(pipe))
```

## F.A.Q.

### Why build on top of ZIP?

ZIP provides several advantages:
- Universally supported file format
- No additional dependencies for reading
- Built-in file indexing
- Wide language support

Why not use a TAR with a table of contents at the beginning of the archive? See explanation [in this comment](https://github.com/huggingface/huggingface_hub/pull/2692#issuecomment-2519863726).

### Why no compression?

- Enables direct memory mapping of large files.
- Ensures consistent and predictable remote file access.
- Prevents CPU overhead during file reading.
- Maintains compatibility with safetensors.

### Can I modify a DDUF file?

> No. For now, DDUF files are designed to be immutable. To update a model, create a new DDUF file.

Reviewer 2: ```suggestion
No. For now, DDUF files are designed to be immutable. To update a model, create a new DDUF file.

### Which all frameworks/ apps support DDUFs?

1. [diffusers](https://github.com/huggingface/diffusers)

We are continuously reaching out to other libraries and frameworks, if you are interested in adding support for your project, open an Issue in [huggingface_hub](https://github.com/huggingface/huggingface_hub/issues) library.
```

I think adding a section like this would be helpful.

# DDUF

<div class="flex justify-center">
    <img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/DDUF/DDUF-Banner.svg"/>
     <img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/DDUF/DDUF-Banner-dark.svg"/>
</div>

## Overview

DDUF (**D**DUF’s **D**iffusion **U**nified **F**ormat) is a file format designed to make storing, distributing, and using diffusion models much easier. Built on the ZIP file format, DDUF offers a standardized, efficient, and flexible way to package all parts of a diffusion model into a single, easy-to-manage file.

This work draws inspiration from the [GGUF](https://github.com/ggerganov/ggml/blob/master/docs/gguf.md) format.

We've seeded some DDUFs of popular formats for the community to play with: https://huggingface.co/DDUF, check them out!

> <Tip>

Reviewer 2: ```suggestion

<Tip>
```

# DDUF

<div class="flex justify-center">
    <img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/DDUF/DDUF-Banner.svg"/>
     <img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/DDUF/DDUF-Banner-dark.svg"/>
</div>

## Overview

DDUF (**D**DUF’s **D**iffusion **U**nified **F**ormat) is a file format designed to make storing, distributing, and using diffusion models much easier. Built on the ZIP file format, DDUF offers a standardized, efficient, and flexible way to package all parts of a diffusion model into a single, easy-to-manage file.

> 

Author: yes definitely :+1:

## Key Features

1. **Single file** packaging.
2. Based on **ZIP file format** to leverage existing tooling.
3. No compression, ensuring **`mmap` compatibility** for fast loading and saving.
4. **HTTP-friendly**: metadata and file structure can be fetched remotely using HTTP Range requests.
5. **Flexible**: each model component is stored in its own directory, following the current `diffusers` structure.
6. **Safe**: uses `safetensors` as weights saving format and prohibits nested directories to prevent ZIP-bombs.

## Technical specifications

Technically, a `.dduf` file **is** a [`.zip` archive](https://en.wikipedia.org/wiki/ZIP_(file_format)). By building on a universally supported file format, we ensure robust tooling already exists. However, some constraints are enforced to meet diffusion models' requirements:

> - Data must be stored uncompressed (flag `0`), allowing mmap-compatibility.

Reviewer 5: Would it make sense to have a one-liner about why map is needed here?

Author: I rephrased it to mention lazy-loading

- Data must be stored using ZIP64 protocol, enabling saving files above 4GB.
- The archive can only contain `.json`, `.safetensors`, `.model` and `.txt` files.
- A `model_index.json` file must be present at the root of the archive. It must contain a key-value mapping with metadata about the model and its components.
- Each component must be stored in its own directory (e.g., `vae/`, `text_encoder/`). Nested files must use UNIX-style path separators (`/`).
- Each directory must correspond to a component in the `model_index.json` index.

> - Each directory must contain a json config file (one of `config.json`, `tokenizer_config.json`, `image_processor.json`).

Reviewer 5: This should also include `scheduler_config.json`. 

For image processors, we should do `preprocessor_config.json` ([example](https://huggingface.co/google/siglip-base-patch16-384/blob/main/preprocessor_config.json)) (~image_processor.json~). Sincere apologies from my side for not being correct in the first place.

```python
(...)

# Export state_dicts one by one from a loaded pipeline
>>> def as_entries(pipe: DiffusionPipeline) -> Generator[Tuple[str, bytes], None, None]:
...     # Build a generator that yields the entries to add to the DDUF file.
...     # The first element of the tuple is the filename in the DDUF archive. The second element is the content of the file.
...     # Entries will be evaluated lazily when the DDUF file is created (only 1 entry is loaded in memory at a time)
...     yield "vae/config.json", pipe.vae.to_json_string().encode()
...     yield "vae/diffusion_pytorch_model.safetensors", safetensors.torch.save(pipe.vae.state_dict())
...     yield "text_encoder/config.json", pipe.text_encoder.config.to_json_string().encode()
...     yield "text_encoder/model.safetensors", safetensors.torch.save(pipe.text_encoder.state_dict())
...     # ... add more entries here

>>> export_entries_as_dduf(dduf_path="my-cool-diffusion-model.dduf", entries=as_entries(pipe))
```

## F.A.Q.

### Why build on top of ZIP?

ZIP provides several advantages:
- Universally supported file format
- No additional dependencies for reading
- Built-in file indexing

> - Wide language support

Reviewer 5: @julien-c should there be any points pertaining to the connection between using ZIP and how that benefits Xet backend?

> Why not use a TAR with a table of contents at the beginning of the archive? See explanation [in this comment](https://github.com/huggingface/huggingface_hub/pull/2692#issuecomment-2519863726).

Reviewer 6: ```suggestion
### Why not use a TAR with a table of contents at the beginning of the archive?

See the explanation in this [comment](https://github.com/huggingface/huggingface_hub/pull/2692#issuecomment-2519863726).
```

### Why no compression?

> - Enables direct memory mapping of large files.
> - Ensures consistent and predictable remote file access.
> - Prevents CPU overhead during file reading.
> - Maintains compatibility with safetensors.

Reviewer 6: To be consistent with the list above :)

```suggestion
- Enables direct memory mapping of large files
- Ensures consistent and predictable remote file access
- Prevents CPU overhead during file reading
- Maintains compatibility with safetensors
```

### Can I modify a DDUF file?

No. For now, DDUF files are designed to be immutable. To update a model, create a new DDUF file.
	
### Which frameworks/apps support DDUFs?
	
- [diffusers](https://github.com/huggingface/diffusers)
	

> We are continuously reaching out to other libraries and frameworks, if you are interested in adding support for your project, open a Discussion in the [DDUF org](https://huggingface.co/spaces/DDUF/README/discussions).

Reviewer 6: ```suggestion
We are constantly reaching out to other libraries and frameworks. If you are interested in adding support to your project, open a Discussion in the [DDUF org](https://huggingface.co/spaces/DDUF/README/discussions).
```

```py
from diffusers import DiffusionPipeline
import torch

pipe = DiffusionPipeline.from_pretrained(
    "DDUF/FLUX.1-dev-DDUF", dduf_file="FLUX.1-dev.dduf", torch_dtype=torch.bfloat16
).to("cuda")
image = pipe(
    "photo a cat holding a sign that says Diffusers", num_inference_steps=50, guidance_scale=3.5
).images[0]
image.save("cat.png")
```

## F.A.Q.

### Why build on top of ZIP?

ZIP provides several advantages:
- Universally supported file format
- No additional dependencies for reading
- Built-in file indexing
- Wide language support

### Why not use a TAR with a table of contents at the beginning of the archive?

See the explanation in this [comment](https://github.com/huggingface/huggingface_hub/pull/2692#issuecomment-2519863726).

### Why no compression?

- Enables direct memory mapping of large files
- Ensures consistent and predictable remote file access
- Prevents CPU overhead during file reading
- Maintains compatibility with safetensors

### Can I modify a DDUF file?

No. For now, DDUF files are designed to be immutable. To update a model, create a new DDUF file.
	
### Which frameworks/apps support DDUFs?
	

> - [Diffusers](https://github.com/huggingface/diffusers)

Reviewer 2: Note: we might have Eric's `diffuse-rs` (candle-based) come in by monday too. So we'll have two supported frameworsk on day-0
